packages/box/src/dfuse-receiver.ts

```diff
@@ -0,0 +1,242 @@
+            if (trx.undo != prev.undo || trx.block.id != prev.block.id) {
+                if (prev.undo) this.storage.undo(prev.block.id);
+                else if (prev.trace) {
+                    const block = { ...prev.block, transactions: [] as any[] };
+                    for (let t of this.unpushedTransactions) {
+                        block.transactions.push({
+                            id: t.trace.id,
+                            actions: t.trace.matchingActions.map((a) => ({
+                                seq: a.seq,
+                                firstReceiver: a.account,
+                                receiver: a.receiver,
+                                name: a.name,
+                                hexData: a.hexData,
+                            })),
+                        });
+                    }
+                    this.storage.pushJsonBlock(
+                        JSON.stringify(block),
+                        prev.irreversibleBlockNum
+                    );
+                }
+                this.unpushedTransactions = [];
+            }
+        }
+        if (trx.trace !== null) this.unpushedTransactions.push(trx);
+    }
+
+    onMessage(message: GraphqlStreamMessage<any>, stream: Stream): void {
+        if (message.type === "data") {
+            const trx: JsonTrx = message.data.searchTransactionsForward;
+            const prev =
+                this.jsonTransactions.length > 0
+                    ? this.jsonTransactions[this.jsonTransactions.length - 1]
+                    : null;
+            logger.info(
+                `${trx.undo ? "undo block" : "recv block"} ${trx.block.num} ${
+                    trx.trace
+                        ? "trx " + trx.trace.id
+                        : "no matching transactions"
+                }`
+            );
+            if (trx.trace || (prev && prev.trace)) {
+                this.jsonTransactions.push(trx);
+                this.pushTrx(trx);
+                if (
+                    this.jsonTransactions.length - this.numSaved > 10 ||
+                    !trx.trace
+                ) {
+                    logger.info(
+                        `save ${dfuseConfig.jsonTrxFile}: ${this.jsonTransactions.length} transactions and undo entries`
+                    );
+                    fs.writeFileSync(
+                        dfuseConfig.jsonTrxFile + ".tmp",
+                        JSON.stringify(this.jsonTransactions)
+                    );
+                    fs.renameSync(
+                        dfuseConfig.jsonTrxFile + ".tmp",
+                        dfuseConfig.jsonTrxFile
+                    );
+                    this.storage.saveState();
+                    this.numSaved = this.jsonTransactions.length;
+                }
+            }
+            stream.mark({
+                cursor: trx.cursor,
+            });
+        }
+        if (message.type === "error")
+            logger.error(JSON.stringify(message, null, 4));
```

Comment:
did you actually get any error during your tests? should we just log and ignore or should we abort the connection and recover somehow?

Reply:
`@dfuse/client` handles reconnection automatically